### PR TITLE
add remove-binding codemod with tests and documentation

### PR DESCRIPTION
**Pull request by Mikek2252**
Ember 2.7 introduced a new deprecation to EmberBinding https://deprecations.emberjs.com/v2.x#toc_ember-binding. This codemod aims to convert all existing uses of `fooBinding: 'path.to.property'` to `foo: alias('path.to.property')`.

The is additional code that checks if `alias` has already been imported, if not it will import it, If there is already an import statement for `@ember/object/computed` a new statement will not be added but the alias specifier will be added to that statement.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,6 +23,9 @@ node ./bin/cli.js <TRANSFORM NAME> path/of/files/ or/some**/*glob.js
 ## Transforms
 
 <!--TRANSFORMS_START-->
+
+
+`[remove-binding](transforms/remove-binding/README.md)` Ember 2.7 introduced a new deprecation to EmberBinding https://deprecations.emberjs.com/v2.x#toc_ember-binding. This codemod aims to convert all existing uses of `fooBinding: 'path.to.property'` to `foo: alias('path.to.property')`.
```
Comment (oscard0m) [inline README.md]:
```suggestion
`[remove-binding](transforms/remove-binding/README.md)` Ember 2.7 introduced a [new deprecation to EmberBinding](https://deprecations.emberjs.com/v2.x#toc_ember-binding). This codemod aims to convert all existing uses of `fooBinding: 'path.to.property'` to `foo: alias('path.to.property')`.
```

File: transforms/remove-binding/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# remove-binding
+Ember 2.7 introduced a new deprecation to EmberBinding https://deprecations.emberjs.com/v2.x#toc_ember-binding. This codemod aims to convert all existing uses of `fooBinding: 'path.to.property'` to `foo: alias('path.to.property')` and if the current file does not have the alias import the following line will also be added `import { alias } from '@ember/object/computed';`
```
Comment (oscard0m) [inline transforms/remove-binding/README.md]:
```suggestion
Ember 2.7 introduced a [new deprecation to EmberBinding](https://deprecations.emberjs.com/v2.x#toc_ember-binding). This codemod aims to convert all existing uses of `fooBinding: 'path.to.property'` to `foo: alias('path.to.property')` and if the current file does not have the alias import the following line will also be added `import { alias } from '@ember/object/computed';`
```

File: transforms/remove-binding/__testfixtures__/classBindings.output.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+let obj = {
+  classBindings: ['class'],
```
Comment (oscard0m) [inline transforms/remove-binding/__testfixtures__/classBindings.output.js:2]: Ember question, Why `classBindings` stays?

File: transforms/remove-binding/__testfixtures__/alreadyHasComputedImport.input.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+import { not } from '@ember/object/computed';
```
Comment (oscard0m) [inline transforms/remove-binding/__testfixtures__/alreadyHasComputedImport.input.js:1]: Found some use cases **not** working for me. I'm not sure if they are out of the scope of this codemod:

### input:
```js
import Computed, { not } from '@ember/object/computed';
let obj = {
  testBinding: 'this.app',
};
```
### output
```js
import Computed, { not, alias } from '@ember/object/computed';
let obj = {
  test: alias('this.app'),
};
```

Comment (oscard0m) [reply transforms/remove-binding/__testfixtures__/alreadyHasComputedImport.input.js:1]: ### input:
```js
import Computed from '@ember/object/computed';
let obj = {
  testBinding: 'this.app',
};
```
### output
```js
import Computed, { alias } from '@ember/object/computed';
let obj = {
  test: alias('this.app'),
};
```

Comment (oscard0m) [reply transforms/remove-binding/__testfixtures__/alreadyHasComputedImport.input.js:1]: Random edge case:
## input:
```js
const alias = "random_variable_alias_used_for_random_reasons";
let obj = {
  testBinding: 'this.app',
};
```

## output:
```js
import { alias as emberAlias } from '@ember/object/computed';
const alias = "random_variable_alias_used_for_random_reasons";
let obj = {
  test: emberAlias('this.app'), // assuring `emberAlias` variable name is available
};
```

File: transforms/remove-binding/index.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  const root = j(file.source);
+
+  const node = root.find(j.ObjectProperty, {
+    key: {
+      type: 'Identifier',
+    },
+  });
+
+  let filteredNodes = node.filter((path) => {
```
Comment (oscard0m) [inline transforms/remove-binding/index.js]: 💅🏽 just a naming suggestion, feel free to resolve this comment.

```suggestion
  let nodesWithBindings = node.filter((path) => {
```

File: transforms/remove-binding/__testfixtures__/alreadyImported.input.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+import { alias } from '@ember/object/computed';
```
Comment (oscard0m) [inline transforms/remove-binding/__testfixtures__/alreadyImported.input.js:1]: Edge case maybe you would like to cover:
## input
```js
import { alias as alias2 } from '@ember/object/computed';
let obj = {
  testBinding: 'this.app',
};
```

## output
```js
import { alias as alias2 } from '@ember/object/computed';
let obj = {
  test: alias2('this.app'),
};
```

File: transforms/remove-binding/index.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    return path.value.key.name && path.value.key.name.endsWith('Binding');
+  });
+
```
Comment (oscard0m) [inline transforms/remove-binding/index.js:17]: 💅🏽 I would early return here to safe some execution time:

```suggestion
if(!filteredNodes.length) {
   return node.toSource();
}
```

File: transforms/remove-binding/index.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    return importt.value.source.value === '@ember/object/computed';
+  });
+
+  let alreadyHasAlias =
+    computedImports.length &&
+    computedImports
+      .get()
+      .node.specifiers.filter((specifier) => specifier.imported.name === 'alias');
+
+  if (!alreadyHasAlias[0] && filteredNodes.length) {
```
Comment (oscard0m) [inline transforms/remove-binding/index.js]: 💅🏽 if we want to treat this as a boolean (from what I see in L30), I would use `Array.some()`

```suggestion
    computedImports
      .get()
      .node.specifiers.some((specifier) => specifier.imported.name === 'alias');

  if (!alreadyHasAlias && filteredNodes.length) {
```

File: transforms/remove-binding/index.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      } else {
+        root.get().node.program.body.unshift(computedImportStatement); // begining of file
+      }
+    }
+  }
+
+  filteredNodes.forEach((path) => {
+    path.value.value = j.callExpression(j.identifier('alias'), [path.value.value]);
+    path.value.key.name = path.value.key.name.slice(0, -7);
+  });
```
Comment (oscard0m) [inline transforms/remove-binding/index.js]: 💅🏽 , just for readability, but feel free to resolve:

```suggestion
  filteredNodes.forEach(({ value: bindingProp }) => {
    bindingProp.value = j.callExpression(j.identifier('alias'), [bindingProp.value]);
    bindingProp.key.name = bindingProp.key.name.slice(0, -7);
  });
```

File: transforms/remove-binding/__testfixtures__/classBindings.output.js (hunk 3 above)
Comment (Mikek2252) [reply transforms/remove-binding/__testfixtures__/classBindings.output.js:2]: So classBindings are still fine to use as it was a way to put classes on the root tag of the component, which you would not have access to in the hbs file. In the latest version of ember they may not be used any more as tag less component have become the norm.

File: transforms/remove-binding/__testfixtures__/alreadyHasComputedImport.input.js (hunk 4 above)
Comment (Mikek2252) [reply transforms/remove-binding/__testfixtures__/alreadyHasComputedImport.input.js:1]: I think i'll skip the last edge case, but am going to add the firsts two tests and they are fairly common cases, thanks!

Comment (oscard0m) [reply transforms/remove-binding/__testfixtures__/alreadyHasComputedImport.input.js:1]: Sure! These are just suggestions. You have more visibility on which cases make sense to cover and which ones do not. :)

**Comment by Mikek2252** [comment]
@oscard0m Hey i've updated the fixtures to include the ones you suggested aside from the two edge cases, i had to update the implementation so you will want to re review.

**Review by oscard0m — APPROVED** [review]
